### PR TITLE
Refactor some code to native promises

### PR DESCRIPTION
**Pull request by Graham42**
Addresses #7298 Part A 1.1

### CODE REVIEW

File: lib/plugins/aws/customResources/index.js [hunk 1]
```diff
@@ -13,191 +13,189 @@ const prepareCustomResourcePackage = _.memoize(zipFilePath =>
...
+    functionName = awsProvider.naming.getCustomResourceApiGatewayAccountCloudWatchRoleHandlerFunctionName();
+    Handler = 'apiGatewayCloudWatchRole/handler.handler';
+    customResourceFunctionLogicalId = awsProvider.naming.getCustomResourceApiGatewayAccountCloudWatchRoleHandlerFunctionLogicalId();
+  } else {
+    throw new serverless.classes.Error(
+      `No implementation found for Custom Resource "${resourceName}"`
+    );
+  }
+  absoluteFunctionName = `${funcPrefix}-${functionName}`;
+  if (absoluteFunctionName.length > 64) {
```
Comment (medikoo) [inline lib/plugins/aws/customResources/index.js:58]: In this method also let's convert all .then() calls to `await` syntax

File: lib/plugins/aws/package/lib/generateCoreTemplate.js [hunk 2]
```diff
@@ -134,6 +132,6 @@ module.exports = {
       this.serverless.service.provider.compiledCloudFormationTemplate
     );
 
-    return BbPromise.resolve();
+    return;
```
Comment (medikoo) [inline lib/plugins/aws/package/lib/generateCoreTemplate.js:135]: This can be removed

File: lib/plugins/package/lib/zipService.js [hunk 3]
```diff
@@ -13,20 +13,15 @@ const _ = require('lodash');
...
       exclude,
       include,
       zipFileName,
-    };
-
-    return BbPromise.bind(this)
-      .then(() => BbPromise.resolve(params))
-      .then(this.excludeDevDependencies)
-      .then(this.zip);
+    }).then(this.zip);
```
Comment (medikoo) [inline lib/plugins/package/lib/zipService.js:21]: Let's convert `.then` to `await` syntax

File: lib/plugins/package/lib/zipService.js [hunk 4]
```diff
@@ -37,17 +32,17 @@ module.exports = {
...
+        params.exclude = _.union(params.exclude, exAndInNode.exclude); //eslint-disable-line
+        params.include = _.union(params.include, exAndInNode.include); //eslint-disable-line
+        return params;
+      } catch (e) {
+        return params;
+      }
     }
 
-    return BbPromise.resolve(params);
+    return params;
```
Comment (medikoo) [inline lib/plugins/package/lib/zipService.js:45]: If I see correctly, a more proper setup of this logic would be as follows:

```javascript
const exAndInNode = await (async () => {
  try { return await excludeNodeDevDependenciesMemoized(servicePath); } catch {}
})();
if (exAndInNode) {
  params.exclude = _.union(params.exclude, exAndInNode.exclude);
  params.include = _.union(params.include, exAndInNode.include); 
}
return params;
```

File: lib/plugins/aws/lib/validate.js [hunk 5]
```diff
@@ -1,19 +1,14 @@
 'use strict';
 
-const BbPromise = require('bluebird');
-
 module.exports = {
-  validate() {
+  async validate() {
```
Comment (medikoo) [inline lib/plugins/aws/lib/validate.js:4]: This one ideally should be kept as _sync_ function, there's no _async_ logic in it, it shouldn't return promise.

**Review by medikoo — CHANGES_REQUESTED** [review]
@Graham42 thank you! Looks very good, I've proposed few minor improvements. Let me know

File: lib/plugins/package/lib/zipService.js (hunk 4 above)
Comment (Graham42) [reply lib/plugins/package/lib/zipService.js:45]: The inline `await` seems a bit harder to read to me (subjective). What about:
```js
      let exAndInNode;
      try {
        exAndInNode = await excludeNodeDevDependenciesMemoized(servicePath);
      } catch {}
      if (exAndInNode) {
        params.exclude = _.union(params.exclude, exAndInNode.exclude); //eslint-disable-line
        params.include = _.union(params.include, exAndInNode.include); //eslint-disable-line
      }
      return params;
```

I can go with whichever you prefer

File: lib/plugins/aws/lib/validate.js (hunk 5 above)
Comment (Graham42) [reply lib/plugins/aws/lib/validate.js]: This caused some cascading test failures and so more changes. I've split the changes for validate into a separate commit to hopefully make it easier to review.

**Comment by Graham42** [comment]
Tests passing now :sweat_smile: so I think it's ready for re-review

File: lib/plugins/package/lib/zipService.js (hunk 4 above)
Comment (medikoo) [reply lib/plugins/package/lib/zipService.js]: It's totally fine! Technically I proposed IIFE simply to use `const` over `let`, but both ways are ok imo (it's purely matter of an opinion)

File: lib/plugins/aws/common/index.js [hunk 6]
```diff
@@ -46,15 +45,13 @@ class AwsCommon {
...
-      'aws:common:cleanupTempDir:cleanup': () => BbPromise.bind(this).then(this.cleanupTempDir),
+      'aws:common:cleanupTempDir:cleanup': async () => this.cleanupTempDir(),
 
-      'aws:common:moveArtifactsToPackage:move': () =>
-        BbPromise.bind(this).then(this.moveArtifactsToPackage),
+      'aws:common:moveArtifactsToPackage:move': async () => this.moveArtifactsToPackage(),
 
-      'aws:common:moveArtifactsToTemp:move': () =>
-        BbPromise.bind(this).then(this.moveArtifactsToTemp),
+      'aws:common:moveArtifactsToTemp:move': async () => this.moveArtifactsToTemp(),
```
Comment (medikoo) [inline lib/plugins/aws/common/index.js:54]: Hooks can be _sync_. I'd leave them not marked as _async_ functions. Let's leave this (whether it's _async_ or _sync_) up to functions that are invoked

File: lib/plugins/package/lib/zipService.js [hunk 7]
```diff
@@ -37,17 +33,18 @@ module.exports = {
...
-        .then(exAndInNode => {
-          params.exclude = _.union(params.exclude, exAndInNode.exclude); //eslint-disable-line
-          params.include = _.union(params.include, exAndInNode.include); //eslint-disable-line
-          return params;
-        })
-        .catch(() => params);
+      let exAndInNode;
+      try {
+        exAndInNode = await excludeNodeDevDependenciesMemoized(servicePath);
+      } catch {} //eslint-disable-line
```
Comment (medikoo) [inline lib/plugins/package/lib/zipService.js:39]: Let's not add eslint disablers (why it's added here?)

**Review by medikoo — CHANGES_REQUESTED** [review]
That looks great @Graham42! I have just few last comments, and we should be good to go

File: lib/plugins/package/lib/zipService.js (hunk 7 above)
Comment (Graham42) [reply lib/plugins/package/lib/zipService.js:39]: I put this try catch in here to replicate the previous behaviour with `.catch()`. I hadn't looked into what the `excludeNodeDevDependenciesMemoized` call was doing.
Looking now, it appears that `excludeNodeDevDependencies` already has a try / catch fallback so we don't need the try catch at this level. I'll remove it.

Comment (medikoo) [reply lib/plugins/package/lib/zipService.js]: My comment was strictly about `// eslint-disable-line`, still if `try/catch` doesn't seem justified indeed it's better to remove it

**Review by medikoo — APPROVED** [review]
Thank you @Graham42, looks great!

**Pull request merged**